File: packages/i18nify-js/src/modules/postalCode/__tests__/getPostalCodeFormat.test.ts
import getPostalCodeFormat from '../getPostalCodeFormat';

describe('getPostalCodeFormat', () => {
  it('returns postal code metadata for India', () => {
    expect(getPostalCodeFormat('IN')).toEqual({
      country_name: 'India',
      format: 'numeric',
      zip_regex: '\\d{6}',
      examples: ['110034', '110001'],
    });
  });

  it('supports lowercase input', () => {
    expect(getPostalCodeFormat('us')).toEqual({
      country_name: 'United States',
      format: 'numeric',
      zip_regex: '(\\d{5})(?:[ \\-](\\d{4}))?',
      examples: ['95014', '22162-1010'],
    });
  });

  it('trims surrounding whitespace', () => {
    expect(getPostalCodeFormat('  HK  ')).toEqual({
      country_name: 'Hong Kong',
      format: 'none',
      zip_regex: '',
      examples: [],
    });
  });

  it('throws for empty input', () => {
    expect(() => getPostalCodeFormat('')).toThrow(
      'countryCode must be a non-empty string.',
    );
  });

  it('throws for unsupported country codes', () => {
    expect(() => getPostalCodeFormat('XX')).toThrow(
      'No postal code data found for country code: "XX".',
    );
  });
});

File: packages/i18nify-go/modules/postal_code/get_postal_code_format_test.go
package postal_code

import (
	"strings"
	"testing"

	"github.com/stretchr/testify/assert"
)

func TestGetPostalCodeFormat_KnownCodes(t *testing.T) {
	tests := []struct {
		cc          string
		countryName string
		format      PostalCodeFormat
		zipRegex    string
		examples    []string
	}{
		{"IN", "India", FormatNumeric, "\\d{6}", []string{"110034", "110001"}},
		{"US", "United States", FormatNumeric, "(\\d{5})(?:[ \\-](\\d{4}))?", []string{"95014", "22162-1010"}},
		{"HK", "Hong Kong", FormatNone, "", []string{}},
	}

	for _, tt := range tests {
		t.Run(tt.cc, func(t *testing.T) {
			got, err := GetPostalCodeFormat(tt.cc)
			assert.NoError(t, err)
			assert.Equal(t, tt.countryName, got.CountryName)
			assert.Equal(t, tt.format, got.Format)
			assert.Equal(t, tt.zipRegex, got.ZipRegex)
			assert.Equal(t, tt.examples, got.Examples)
		})
	}
}

func TestGetPostalCodeFormat_LowercaseInput(t *testing.T) {
	got, err := GetPostalCodeFormat("in")
	assert.NoError(t, err)
	assert.Equal(t, "India", got.CountryName)
	assert.Equal(t, FormatNumeric, got.Format)
}

func TestGetPostalCodeFormat_WithWhitespace(t *testing.T) {
	got, err := GetPostalCodeFormat("  HK  ")
	assert.NoError(t, err)
	assert.Equal(t, "Hong Kong", got.CountryName)
	assert.Equal(t, FormatNone, got.Format)
}

func TestGetPostalCodeFormat_InvalidCode(t *testing.T) {
	_, err := GetPostalCodeFormat("XX")
	assert.Error(t, err)
	assert.True(t, strings.Contains(err.Error(), "no postal code data"))
}

func TestGetPostalCodeFormat_EmptyCode(t *testing.T) {
	_, err := GetPostalCodeFormat("")
	assert.Error(t, err)
	assert.True(t, strings.Contains(err.Error(), "non-empty string"))
}
